File: Data_Structure/DS_Using_Python/Perfect_Binary_Tree.py
""" Construct below tree
             1
           /   \
          /     \
         2       3
        / \     / \
       4   5   6   7
    """
class PerfectBinaryTree:
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None
		
	#Calculate the depth
	def depth(self, node):
		d = 0
		while node is not None:
			d = d+1
			node = node.left
		return d
		
	#Check if the tree is perfect
	def isPerfectBinaryTree(self, tree, d, level):
		if tree is None:
			return True
		if tree.left is None and tree.right is None:
			d  = level+1
			return d
		if tree.left is None or tree.right is None:
			return False
		return self.isPerfectBinaryTree(tree.left, d ,level+1) and self.isPerfectBinaryTree(tree.right, d, level+1)
		
	def isPerfect(self, root):
		d = self.depth(root)
		return self.isPerfectBinaryTree(root, d, 0)
	
	#insert data in left child node
	def insertLeft(self, newNode):
		if self.left is None:
			self.left = PerfectBinaryTree(newNode)
		else:
			tree = PerfectBinaryTree(newNode)
			tree.left = self.left
			self.left = tree
		
	#insert data in right child node
	def insertRight(self, newNode):
		if self.right is None:
			self.right = PerfectBinaryTree(newNode)
		else:
			tree = PerfectBinaryTree(newNode)
			tree.right = self.right
			self.right = tree
		
root = PerfectBinaryTree(1)
root.insertLeft(2)
root.insertRight(3)
root.insertLeft(4)
root.insertRight(5)
root.insertLeft(6)
root.insertRight(7)

if(root.isPerfect(root)):
	print("It is a Perfect Binary Tree")
else:
	print("It is not a Perfect Binary Tree")

File: Data_Structure/Trees/Perfect_Binary_Tree.c
#include<stdio.h>
#include<stdbool.h>

struct node
{
    int data;
    struct node* left;
    struct node* right;
};

//Creating a new node
struct node* createNode(int item)
{
    struct node* newNode = (struct node*)malloc(sizeof(struct node));
    newNode -> data = item;
    newNode -> left = NULL;
    newNode -> right = NULL;
    return newNode;
}

//Calculate the depth
int depth(struct node* node)
{
    int i = 0;
    while(node != NULL)
    {
        i += 1;
        node = node -> left;
    }
    return i;
}

//Check if the tree is perfect
bool isPerfectBinaryTree(struct node* root, int d, int level)
{
    if(root == NULL)
    {
        return true;
    }
    if(root -> left == NULL && root -> right == NULL)
    {
        return d = level+1;
    }
    if(root -> left == NULL || root -> right == NULL)
    {
        return false;
    }
    
    return(isPerfectBinaryTree(root -> left, d, level+1) && isPerfectBinaryTree(root -> right, d, level+1));
}

//Insert data in left child node
struct node* insertLeft(struct node* root, int item)
{
    root -> left = createNode(item);
    return root -> left;
}

//Insert data in right child node
struct node* insertRight(struct node* root, int item)
{
    root -> right = createNode(item);
    return root -> right;
}

//Wrapper function 
bool isPerfect(struct node *root)
{ 
    int d = depth(root); 
    return isPerfectBinaryTree(root, d, 0); 
}

int main()
{
    struct node* root = createNode(2);
    insertLeft(root, 5);
    insertRight(root, 7);
    insertLeft(root -> left, 8);
    insertRight(root -> left, 82);
    insertLeft(root -> right, 10);
    insertRight(root -> right, 29);
    
    if(isPerfect(root))
    {
        printf("It is a Perfect Binary Tree");
    }
    else
    {
        printf("It is not a Perfect Binary Tree");
    }
    
    return 0;
}